torch/_expanded_weights/expanded_weights_impl.py

```diff
@@ -0,0 +1,203 @@
+import torch
+import functools
+import warnings
+
+from typing import Callable, Dict, List, Union
+
+HANDLED_FUNCTIONS: Dict[Callable, torch.autograd.Function] = {}
+
+def implements_per_sample_grads(torch_function):
+    @functools.wraps(torch_function)
+    def decorator(autograd_func):
+        HANDLED_FUNCTIONS[torch_function] = autograd_func
+        return autograd_func
+    return decorator
+
+# ExpandedWeight represents a weight (parameter) Tensor that has an expanded
+# batch dimension. Operations on the ExpandedWeight Tensor take advantage of
+# how the batch dimension is expanded by de-expanding the weight before
+# computation. A subsequent call to .backward() computes gradients for
+# ExpandedWeight. Those gradients are equivalent to per-sample-grads for the
+# unexpanded weight Tensors.
+#
+# ExpandedWeight has a fallback that does the forward + backward computation.
+# The backward computation is not optimized: it runs torch.autograd.grad in
+# a loop. To optimize the backward computation further, we must register
+# overrides for specific operators.
+#
+# This is a __torch_function__ object but it could have also been a Tensor Extension
+# with a dispatch key.
+class ExpandedWeight(torch.Tensor):
+    handled_functions = HANDLED_FUNCTIONS
+
+    # needed for conv2d default kwargs
+    conv_kwarg_options = ['stride', 'padding', 'dilation', 'groups']
+    conv_kwarg_defaults = {'stride': 1, 'padding': 0, 'dilation': 1, 'groups': 1}
+
+    def __new__(cls, orig_weight, batch_size):
+        ret = torch.Tensor._make_subclass(cls, orig_weight.detach(), orig_weight.requires_grad)
+        if not isinstance(orig_weight, torch.Tensor):
+            raise RuntimeError(f"Can only make ExpandedWeights of Tensors, got {type(orig_weight).__name__}")
+        cls.batch_size = batch_size
+        cls.orig_weight = orig_weight
+        return ret
+
+    @classmethod
+    def __torch_function__(cls, func, _, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+        if func not in cls.handled_functions:
+            warnings.warn(f"don't have custom implementation for function {func.__name__}. Using slow fallback")
+            return SlowFallback.apply(func, *(args + tuple(kwargs.values())))
+        if func == torch.nn.functional.conv2d:
+            remaining_kwargs = 7 - len(args)
+            remaining_kwargs_options = cls.conv_kwarg_options[4 - remaining_kwargs:]
+            kwargs = {key: cls.conv_kwarg_defaults[key] for key in remaining_kwargs_options} | kwargs
+        return cls.handled_functions[func].apply(*(args + tuple(kwargs.values())))
+
+    @property
+    def shape(self):
+        return self.orig_weight.shape
```

Comment:
What happens if you call `expanded_weight.size()`? does that return the correct thing?

Reply:
Added `size()` function

Reply:
Follow-up question: (maybe more for @albanD) is user error to pass a `__torch_function__` tensor subclass into some C++ code that requires a Tensor? For the PyTorch frontend API the answer to this is probably no because it should enter `__torch_function__` but if users have things like custom C++ operators that they've pybind'ed into Python, do we consider this to be a user error?

Reply:
It is not an error and the code will use the underlying Tensor as-is. This is how `nn.Parameter` works today.
But it is a subtlety that the user needs to be aware, as soon as you pass the frontend API binding or enter any other c++ function. Only the c++ Tensor associated with your subclass will exist.